azext_iot/digitaltwins/providers/model.py

```diff
@@ -76,10 +88,61 @@ def add(self, models=None, from_directory=None):
+                    model_id_to_model_map[model_def['@id']] = model_def
+
+                # Create a dictionary to categorize models by their number of dependencies
+                dep_count_to_models_map = {}
+                for model in payload:
+                    num_dependencies = len(get_model_dependencies(model, model_id_to_model_map))
+                    if num_dependencies not in dep_count_to_models_map:
+                        dep_count_to_models_map[num_dependencies] = []
+                    dep_count_to_models_map[num_dependencies].append(model)
+
+                # Sort the dictionary by dependency count
+                dep_count_to_models_map = dict(sorted(dep_count_to_models_map.items()))
+                models_batch = []
+                response = []
+                pbar = tqdm(total=len(payload), desc='Creating models...', ascii=' #')
+                # The map being iterated is sorted by dependency count, hence models with 0 dependencies go first,
+                # followed by models with 1 dependency, then 2 dependencies and so on... This ensures that all dependencies
+                # of each model being added were either already added in a previous iteration or are in the current payload.
+                for models_list in dep_count_to_models_map.values():
+                    while len(models_batch) + len(models_list) > MAX_MODELS_PER_BATCH:
+                        num_models_to_add = MAX_MODELS_PER_BATCH - len(models_batch)
+                        models_batch.extend(models_list[0:num_models_to_add])
+                        response.extend(self.model_sdk.add(models_batch, raw=True).response.json())
+                        models_created.extend([model['@id'] for model in models_batch])
+                        pbar.update(len(models_batch))
+                        # Remove the model ids which have been processed
+                        models_list = models_list[num_models_to_add:]
+                        models_batch = []
+                    models_batch.extend(models_list)
+                # Process the last set of model ids
+                if len(models_batch) > 0:
+                    pbar.update(len(models_batch))
+                    response.extend(self.model_sdk.add(models_batch, raw=True).response.json())
+                pbar.close()
+                return response
             return self.model_sdk.add(payload, raw=True).response.json()
         except ErrorResponseException as e:
+            if len(models_created) > 0:
+                pbar.close()
+                logger.error(
+                    "Error creating models. Deleting {} models already created by this operation...".format(len(models_created))
+                )
+                # In case of error delete all models created by this operation. Models will be deleted in the reverse
+                # order they were created, hence ensuring each model's dependencies are deleted after deleting the model.
```

Comment:
this is nice - we could just say "you clean up your mess" but you go the extra step :D. 

Would there be an issue in this scenario:
person has a model A already in the dt instance. Person tries to add a bunch of models, with model A in that list (and the add fails). Will model A be deleted from the dt instance?

Reply:
The logic only deletes the models in previously successful batches. Hence, in this example model "A" doesn't get deleted since it was not successfully created as a part of this command run.

Reply:
So it doesnt touch anything in the unsuccessful batches - would anything from the unsuccessful batches be possibly added to the dt instance? (in the scenario, lets say that B is part of the batch with A. Is there a chance that B got added in?)

Reply:
triple checking that a bad model in a batch makes the service reject the entire batch

Reply:
Correct the entire batch containing the bad model is rejected. None of the models from an unsuccessful batch make it to the DT instance.